activerecord/test/cases/associations/has_one_associations_test.rb

```diff
@@ -690,6 +690,27 @@ def test_has_one_assignment_triggers_save_on_change_on_replacing_object
     assert_equal "new name", pirate.ship.reload.name
   end
 
+  def test_has_one_assignment_triggers_save_when_belongs_to_already_assigned
+    pirate = Pirate.create!(catchphrase: "Why is the rum always gone?")
+    original_ship = Ship.create!(pirate: pirate, name: "Original Ship")
+
+    assert_equal "Original Ship", pirate.reload.ship.name
+
+    new_ship = Ship.new(name: "New Ship")
+    new_ship.pirate = pirate
+    assert_queries(3) do
+      # One query each to:
+      # - select the old ship (reload)
+      # - nullify the old ship
+      # - update the new ship
+      pirate.ship = new_ship
```

Comment:
The downside here is we end up with 1 additional query caused by the reload.

Reply:
In the case of a double assignment,  naturally the in-memory version of was pulled from the database already has been discarded. It makes sense to me that there will be have to be a reload of that entity. I would imagine we don't want the reload to happen in the generalised case though but that doesn't seem to be the case here - the reload is only when necessary - the least surprising or more correct behaviour of the library is the desired outcome.